File: 0001 TwoSum/TwoSum.cs
public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        // Brute Force
        // for (int i = 0; i < nums.length - 1; i++)
        // for (int j = i + 1; j < nums.length; j++)
        //     if (nums[i] + nums[j] == target)
        //         return new int[2] {i, j};
        // return null;

        // Two-pass Hash Table
        // Dictionary<int, int> dic = new Dictionary<int, int>();
        // for (int i = 0; i < nums.Length; i++)
        //     if (!dic.ContainsKey(target - nums[i]))
        //         dic.Add(target - nums[i], i);

        // for (int i = 0; i < nums.Length - 1; i++)
        //     if (dic.ContainsKey(nums[i]))
        //         return new int[2] {i, dic[nums[i]]};
        // return null;

        // One-pass Hash Table
        Dictionary<int, int> dic = new Dictionary<int, int>();
        int key;
        for (int i = 0; i < nums.Length; i++)
            if (dic.ContainsKey(nums[i]))
                return new int[] {dic[nums[i]], i};
            else
            {
                key = target - nums[i];
                if (!dic.ContainsKey(key))
                    dic.Add(key, i);
            }
        return null;
    }
}

File: 0001 TwoSum/TwoSum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Brute Force
        # for i in range(0, len(nums) - 1):
        #     for j in range(i + 1, len(nums)):
        #         if (nums[i] + nums[j] == target):
        #             return [i, j]
        # assert False

        # Two-pass Hash Table
        # dic = dict()
        # for i in range(len(nums)):
        #     dic[target - nums[i]] = i

        # for i in range(0, len(nums) - 1):
        #     if (nums[i] in dic):
        #         return [i, dic[nums[i]]]
        # assert False

        # One-pass Hash Table
        dic = dict()
        for i, n in enumerate(nums):
            if n in dic:
                return dic[n], i
            dic[target - n] = i
        assert False
